doc/design/eventdriven-callbacks.md

```diff
@@ -0,0 +1,215 @@
+  myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED, sampleReceived);
+  myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::UNSUBSCRIBED, waitForSubscription);
+  myCallbackReactal.detachEvent(subscriber, iox::popo::SubscriberEvent::SUBSCRIBED);
+}
+
+void waitForUnsubscribe(iox::popo::UntypedSubscriber & subscriber) {
+  std::cout << "unsubscribed from publisher\n";
+  myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::SUBSCRIBED, waitForSubscription);
+  myCallbackReactal.detachEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED);
+  myCallbackReactal.detachEvent(subscriber, iox::popo::SubscriberEvent::UNSUBSCRIBED);
+}
+
+int main() {
+  myCallbackReactal.attachEvent(mySubscriber, iox::popo::SubscriberEvent::SUBSCRIBED, waitForSubscription);
+}
+
+```
+
+### Solution
+#### Condition Variable
+
+  - Create from `ConditionVariableWithOriginTracing` (yeah we need a better name) has `ConditionVariableData` 
+    as a member and adds: `std::atomic_bool m_triggeredBy[MAX_CALLBACKS];`
+
+    **Reason:**
+      - The reactal will iterate through this array to find out by which event it was triggered 
+      - It is cache local, since it is one piece of contiguous memory which can be put directly 
+        into the CPU cache when it is being iterated. Therefore it is very performant.
+      - WaitSet approach are callbacks which have an overhead implementation 
+        and performance wise.
+        * implementation, since we have to manage set and reset all the callbacks 
+        * performance, when we call a callback with every iteration, we have to first load the 
+            callback into the cpu cache (generates cache misses for all the other callback 
+            elements in the vector which is being iterated). In the callback a method in an 
+            object is called (the object has to be loaded again into the cpu cache, cache miss for 
+            every member which is not used by the method).
+
+    **Adjustments:**
+      - Create the class with a template bool array size argument to be more flexible and memory efficient.
+
+    **Note:**
+      - Sadly, we cannot pursue this approach in the WaitSet since it is event 
+        and state based at the same time and this approach supports only event 
+        driven triggering.
+
+  - add class `ConditionVariableWithOriginTracingSignaler` (I know, better name)
+  
+    **Reason:** it behaves quiete differently then the `ConditionVariable`. The `ConditionVariable` tells you:
+    someone has signalled you and you have to find out who it was. The `ConditionVariableWithOriginTracing`
+    would tell you: A, B and C have signalled you.
+
+    ```cpp
+    class ConditionVariableWithOriginTracingSignaler {
+      public:
+        // id = corresponds to the id in the ConditionVariable atomic_bool array to 
+        // identify the trigger origin easily
+        void notifyOne(const uint64_t id); 
```

Comment:
NotifySpecifc?

Reply:
Misunderstood it: id = who is notifying

Reply:
Every instance which can signal an event has its own id and signals the event directly. There will most likely never be some kind of class which collects events and notifies a set of ids.